### ISSUE TITLE
Assign IP/MAC/Domain to license keys

### ISSUE THREAD

**Issue by drazenbebic** [drazenbebic/license-manager-for-woocommerce#21]
Add a new text parameter to the license keys. This parameter will hold an IP/MAC address or domain. When the license key is being activated, there must be a check to match the data from the license key with the data from the request. If these do not match, then the license key cannot be activated.

**Comment by daveboulden** [comment]
This is the other enhancement I would very much like to see. It would be good if when specifying a domain, a list of domains with wildcards could be handled. 

e.g. `*.mysite.com, www.myothersite.com` for all subdomains of mysite.com and www.myothersite.com.

**Comment by drazenbebic** [comment]
@srcmkr @stijnb1234 @sebastienserre @maxi07 @otakuphp @daveboulden 

I need some help with conceptualizing this feature. From what I understand, it should work like this.

1.  Client sends activation request to sender.
2. The server retrieves the domain from the request (**From where? Is it contained in the request body?**)
3. The server tries to match the retrieved license against a list of allowed domains (**Defined by the administrator on a product level? on a generator level?**)
4. If the licensing server is unable the match, the activation request cannot proceed.

Please let me know what you guys think.

**Comment by daveboulden** [comment]
My feedback would be:

1. Correct
2. The originating domain should be in the 'host' value in the HTTP header or 'origin' in a CORS HTTP header. The 'host' value is optional for HTTP 1.0, but mandatory from HTTP 1.1, but putting it in the body of the request leaves it open to mis-use if the REST request is intercepted and the request body data changed to a fixed domain name.
3. I would say the domain is set at purchased product level i.e. the buyer chooses the domain to use the license with when they purchase it.
4. Correct

**Comment by drazenbebic** [comment]
@daveboulden 

Regarding 2. - What about activation requests from devices? For example when developing an android/ios application which will run on any device. What's the domain in this case? The domain feature needs to be able to be turned on/off on a product level in this case?

Regarding 3. - Do you mean that the buyer, the person who purchases the license, can type in the domain from which the license will be used in one of the above-mentioned formats?

**Comment by daveboulden** [comment]
@drazenbebic 
Yes, the domain restriction definitely needs to be optional. I guess it could be optional at the generator level so that I could use different generators for products which should be restricted to a domain and those that should not.

In my scenario, I would selling a commercial WordPress plugin on my WooCommerce site. When someone purchases the plugin, they have to choose the domain the plugin will be licensed for as I would want to be able to set the price for a license to be "per website". The request I made for wildcards or a list of domains to be handled would allow me to sell a license to cover a collection of domains within one purchase so that a customer could buy a license that covered all of their sub-domains for example.

**Comment by stijnb1234** [comment]
For me:

1. Correct
2. You can get the IP or hostname from the sender by the request.

I've done this myself:
```php
$ip = $_SERVER['HTTP_CLIENT_IP'] ? $_SERVER['HTTP_CLIENT_IP'] : ($_SERVER['HTTP_X_FORWARDED_FOR'] ? $_SERVER['HTTP_X_FORWARDED_FOR'] : $_SERVER['REMOTE_ADDR']);
$hostname = gethostbyaddr($ip);

lmfwc_add_license_meta($licenseId, "ip", $ip);
lmfwc_add_license_meta($licenseId, "port", $instance);

if ($hostname != false) {
	lmfwc_add_license_meta($licenseId, "hostname", $hostname);
}
```
It listens to the lmfwc_rest_api_validation filter.

3. Yes. For me, it would be great if the IPs allowed are the same as how many times you can activate it.

4. Correct

**Comment by maxi07** [comment]
But don't forget about Desktop apps, they might send their activation request to your API via _curl_ or as _webrequests_ which catch HTTP status codes or JSON responses.
You could think of using hardware ids to identify machines and handle per machine activations.

![license](https://user-images.githubusercontent.com/7480270/101388003-a5699080-38bf-11eb-8615-b543d6911989.png)
